lib/schema/src/validate.rs

```diff
@@ -751,3 +904,304 @@ impl RegisterBlock {
+                        array_dimensions: vec![3,],
+                        ..Default::default()
+                    }
+                    .into(),
+                ],
+                ..Default::default()
+            }
+        );
+    }
+
+    #[test]
+    fn test_extract_array_multidimensional() {
+        let block = RegisterBlock {
+            name: "block".into(),
+            registers: vec![
+                Register {
+                    name: "unrelated0".into(),
+                    offset: 0,
+                    ..Default::default()
+                }
+                .into(),
+                Register {
+                    name: "data0".into(),
+                    offset: 16,
+                    array_dimensions: vec![16],
+                    ..Default::default()
+                }
+                .into(),
+                Register {
+                    name: "data1".into(),
+                    offset: 20,
+                    array_dimensions: vec![16],
+                    ..Default::default()
+                }
+                .into(),
+                Register {
+                    name: "data2".into(),
+                    offset: 24,
+                    array_dimensions: vec![16],
+                    ..Default::default()
+                }
+                .into(),
+            ],
+            ..Default::default()
+        };
+        let mut block = block.validate_and_dedup().unwrap();
+        block.extract_array("data").unwrap();
+        assert_eq!(
+            block.block,
+            RegisterBlock {
+                name: "block".into(),
+                registers: vec![
+                    Register {
+                        name: "unrelated0".into(),
+                        offset: 0,
+                        ..Default::default()
+                    }
+                    .into(),
+                    Register {
+                        name: "data".into(),
+                        offset: 16,
+                        array_dimensions: vec![3, 16,],
+                        ..Default::default()
+                    }
+                    .into(),
+                ],
+                ..Default::default()
+            }
+        );
+    }
+
+    #[test]
+    fn test_extract_array_dimension_mismatch() {
+        let block = RegisterBlock {
+            name: "block".into(),
+            registers: vec![
+                Register {
+                    name: "data0".into(),
+                    offset: 16,
+                    ..Default::default()
+                }
+                .into(),
+                Register {
+                    name: "data1".into(),
+                    offset: 20,
+                    array_dimensions: vec![2],
+                    ..Default::default()
+                }
+                .into(),
+            ],
+            ..Default::default()
+        };
+        let mut block = block.validate_and_dedup().unwrap();
+        assert_eq!(
```

Comment:
Any reason not to directly `assert_eq!` the `ValidationError` instead of the string it produces?

Reply:
I can't imagine anyone is ever going to match on these explicit error types, so I thought it was more important that we test that the human-readable string is correct.

Reply:
Ack.